From 956f82528e9a79c77abacc8e20b88bebf5f73912 Mon Sep 17 00:00:00 2001
From: kelson8 <kelson@kelsoncraft.net>
Date: Thu, 17 Sep 2026 18:34:24 -0400
Subject: [PATCH 169/176] feat: Add/fix things in lua

* Make switch_roads_off and switch_ped_roads_off work in lua.

* Added getFadingStatus, drawSphere, and switchRubbish functions into lua_world.

* Add test for drawing sphere to the world in lua, this works but I need to add a toggle in lua.

* Add test function for validating the the road positions for turning them on/off.

* Fix infinite ammo in vehicles.

* Update changelog.
---
 Changelog.md                            |  16 ++
 src/extras/functions/misc_functions.cpp |  27 +++
 src/extras/functions/misc_functions.h   |   1 +
 src/extras/lua/lua_world.cpp            | 307 +++++++++++++++++-------
 src/extras/lua/lua_world.h              |   4 +
 src/weapons/Weapon.cpp                  |  15 +-
 6 files changed, 278 insertions(+), 92 deletions(-)

diff --git a/Changelog.md b/Changelog.md
index 501ab303..f33199ac 100644
--- a/Changelog.md
+++ b/Changelog.md
@@ -1088,3 +1088,19 @@ fix: Make toggle player controls work in lua
 * Add SetControl function to player_functions.
 
 * Update changelog.
+
+**Commit 3**
+
+feat: Add/fix things in lua
+
+* Make switch_roads_off and switch_ped_roads_off work in lua.
+
+* Added getFadingStatus, drawSphere, and switchRubbish functions into lua_world.
+
+* Add test for drawing sphere to the world in lua, this works but I need to add a toggle in lua.
+
+* Add test function for validating the the road positions for turning them on/off.
+
+* Fix infinite ammo in vehicles.
+
+* Update changelog.
diff --git a/src/extras/functions/misc_functions.cpp b/src/extras/functions/misc_functions.cpp
index 4d0984db..798995c5 100644
--- a/src/extras/functions/misc_functions.cpp
+++ b/src/extras/functions/misc_functions.cpp
@@ -165,6 +165,33 @@ MiscFunctions::SwitchPedRoadsOff(float x1, float x2, float y1, float y2, float z
 	ThePaths.SwitchPedRoadsOffInArea(x1, x2, y1, y2, z1, z2, true);
 }
 
+/**
+ * This is a replication of what SWITCH_ROADS_OFF and the other road functions use.
+ * It sets some values to defaults if they are greater then some of the X, Y, and Z values.
+ * TODO Figure out how to make this into a function for toggling the roads.
+ */
+void
+MiscFunctions::SwitchRoadsValidate(float infX, float infY, float infZ, float supX, float supY, float supZ)
+{
+	if(infX > supX)
+	{
+		infX = supX;
+		supX = infX;
+	}
+
+	if (infY > supY)
+	{
+		infY = supY;
+		supY = infY;
+	}
+
+	if(infZ > supZ)
+	{
+		infZ = supZ;
+		supZ = infZ;
+	}
+}
+
 void
 SetZoneCarInfo()
 {
diff --git a/src/extras/functions/misc_functions.h b/src/extras/functions/misc_functions.h
index b5fbf369..d76c5551 100644
--- a/src/extras/functions/misc_functions.h
+++ b/src/extras/functions/misc_functions.h
@@ -86,6 +86,7 @@ public:
 	// Ped road toggling, this toggles the peds in the specified area like in the scripts.
 	void SwitchPedRoadsOn(float x1, float x2, float y1, float y2, float z1, float z2);
 	void SwitchPedRoadsOff(float x1, float x2, float y1, float y2, float z1, float z2);
+	void SwitchRoadsValidate(float infX, float infY, float infZ, float supX, float supY, float supZ);
 	
 	// Objects
 	// Create object, not implemented yet
diff --git a/src/extras/lua/lua_world.cpp b/src/extras/lua/lua_world.cpp
index 4edd5827..1976ef77 100644
--- a/src/extras/lua/lua_world.cpp
+++ b/src/extras/lua/lua_world.cpp
@@ -33,10 +33,13 @@
 #include "Camera.h"
 #include "Streaming.h"
 #include "Pools.h"
+#include "Rubbish.h"
 //#include "Hud.h"
 //#include "Radar.h"
 //#include "Text.h"
 
+#include "SpecialFX.h"
+
 //#include "Automobile.h"
 //#include "Bike.h"
 //#include "Boat.h"
@@ -164,10 +167,18 @@ extern "C"
 		
 		//
 
+		//----------
+		// Fading
+		//----------
 		// Fade the camera
 		lua_pushcfunction(L, fadeCameraLua);
 		lua_setfield(L, -2, "fade_camera");
 
+		// Check if the camera is fading in or out.
+		lua_pushcfunction(L, getFadingStatusLua);
+		lua_setfield(L, -2, "get_fading_status");
+		//
+
 		// Setup the target marker
 		lua_pushcfunction(L, setupTargetMarkerLua);
 		lua_setfield(L, -2, "set_marker");
@@ -197,6 +208,13 @@ extern "C"
 		lua_setfield(L, -2, "set_zone_group");
 		//
 
+		// Draw a sphere on the map
+		lua_pushcfunction(L, drawSphereLua);
+		lua_setfield(L, -2, "draw_sphere");
+
+		// Switch the rubbish on/off on the map
+		lua_pushcfunction(L, switchRubbishLua);
+		lua_setfield(L, -2, "switch_rubbish");
 
 		// World namespace
     	lua_setglobal(L, "world");
@@ -366,63 +384,52 @@ extern "C"
 	 * 
 	 * Usage (Takes two vectors for a leftBottom, and rightTop values):
 	 * https://library.sannybuilder.com/#/vc/script/extensions/default/01E8
-	 * world.switch_roads_off({x = 25, y = 25, z = 25}, {x = 50, y = 50, z = 50})
+	 * world.switch_roads_off(25, 25, 25, 50, 50, 50)
 	 */
 	int switchRoadsOffLua(lua_State *L)
 	{
-		auto leftBottom = luaGetCVector(L, 1);
-		auto rightTop = luaGetCVector(L, 2);
-	
-		if (!leftBottom || !rightTop)
-    	{
-    		return luaL_error(L, "world.switch_roads_off expects two vector tables with x, y and z values." );
+		// auto leftBottom = luaGetCVector(L, 1);
+		// auto rightTop = luaGetCVector(L, 2);
+
+		// Replicated setup from COMMAND_SWITCH_ROADS_OFF.
+		// This works now! I had to change this I guess.
+		float infX = (float)lua_tonumber(L, 1);
+		float infY = (float)lua_tonumber(L, 2);
+		float infZ = (float)lua_tonumber(L, 3);
+		float supX = (float)lua_tonumber(L, 4);
+		float supY = (float)lua_tonumber(L, 5);
+		float supZ = (float)lua_tonumber(L, 6);
+
+		// Adapted from scripts, hopefully this is right..
+		if(infX > supX)
+		{
+			infX = supX;
+			supX = infX;
 		}
-		// ThePaths.SwitchPedRoadsOffInArea(leftBottomX, rightTopX, leftBottomY, rightTopY, leftBottomZ, rightTopZ, true);
 
-		// I modifed this function to be more in line with SWITCH_PED_ROADS_OFF.
-		// If this works, then I'll switch the others to using this format also.
-		// Well this didn't work..
-		// I'm not even sure how this function works honestly, I might just make another one that does this.
-		ThePaths.SwitchRoadsOffInArea(
-			leftBottom->x, 
-			rightTop->x,
-			leftBottom->y,
-			rightTop->y,
-			leftBottom->z, 
-			rightTop->z,
-			true);
-
-		// ThePaths.SwitchRoadsOffInArea(
-		// 	leftBottom->x, 
-		// 	leftBottom->y, 
-		// 	leftBottom->z,			
-		// 	rightTop->x, 
-		// 	rightTop->y, 
-		// 	rightTop->z, 
-		// 	true);
-
-		// For debugging
-		// Well this displays a message, why doesn't it turn off the paths for these roads?
-		std::string roadsMsg = "Vehicle roads have been switched off for: "
-		+ std::string("leftBottomX: ") + std::to_string(leftBottom->x) 
-		+ " leftBottomY: " + std::to_string(leftBottom->y)
-		+ " leftBottomZ: " + std::to_string(leftBottom->z)
-		
-		+ " rightTopX: " + std::to_string(rightTop->x)
-		+ " rightTopY: " + std::to_string(rightTop->y)
-		+ " rightTopZ: " + std::to_string(rightTop->z);
+		if (infY > supY)
+		{
+			infY = supY;
+			supY = infY;
+		}
 
-		// LogFunctions::LogInfo(Defines::luaLogFile, roadsMsg);
+		if(infZ > supZ)
+		{
+			infZ = supZ;
+			supZ = infZ;
+		}
+		// New
+		ThePaths.SwitchRoadsOffInArea(infX, supX, infY, supY, infZ, supZ, true);
 		//
 
-
-		// SwitchRoadsOffInArea
 		return 0;
 	}
 
 	/**
 	 * Switch the vehicle roads on.
 	 * 
+	 * TODO Fix this, match up with turn off function later.
+	 * 
 	 * Usage (Takes two vectors for a leftBottom, and rightTop values):
 	 * https://library.sannybuilder.com/#/vc/script/extensions/default/01E7
 	 * world.switch_roads_on({x = 25, y = 25, z = 25}, {x = 50, y = 50, z = 50})
@@ -451,50 +458,51 @@ extern "C"
 	/**
 	 * Switch the ped roads off in the area.
 	 * 
+	 * This works now!
+	 * TODO Replicate changes for switching both the roads on, shouldn't be hard.
+	 * 
 	 * Usage:
-	 * world.switch_ped_roads_off({x = 25, y = 25, z = 25}, {x = 50, y = 50, z = 50})
+	 * world.switch_ped_roads_off(25, 25, 25, 50, 50, 50)
 	 */
 	int switchPedRoadsOffLua(lua_State *L)
 	{
-		auto leftBottom = luaGetCVector(L, 1);
-		auto rightTop = luaGetCVector(L, 2);
-	
-		if (!leftBottom || !rightTop)
-    	{
-    		return luaL_error(L, "world.switch_ped_roads_off expects two vector tables with x, y and z values." );
+		// auto leftBottom = luaGetCVector(L, 1);
+		// auto rightTop = luaGetCVector(L, 2);
+
+		// Replicated setup from COMMAND_SWITCH_ROADS_OFF.
+		// This works now! I had to change this I guess.
+		float infX = (float)lua_tonumber(L, 1);
+		float infY = (float)lua_tonumber(L, 2);
+		float infZ = (float)lua_tonumber(L, 3);
+		float supX = (float)lua_tonumber(L, 4);
+		float supY = (float)lua_tonumber(L, 5);
+		float supZ = (float)lua_tonumber(L, 6);
+
+		// Adapted from scripts, hopefully this is right..
+		if(infX > supX)
+		{
+			infX = supX;
+			supX = infX;
 		}
 
-		// ThePaths.SwitchPedRoadsOffInArea(
-		// 	leftBottom->x, 
-		// 	rightTop->x,
-		// 	leftBottom->y,
-		// 	rightTop->y,
-		// 	leftBottom->z, 
-		// 	rightTop->z,
-		// 	true);
+		if (infY > supY)
+		{
+			infY = supY;
+			supY = infY;
+		}
 
-		ThePaths.SwitchPedRoadsOffInArea(
-			leftBottom->x, 
-			leftBottom->y, 
-			leftBottom->z,			
-			rightTop->x, 
-			rightTop->y, 
-			rightTop->z, 
-			true);
-
-		// For debugging
-		// Well this displays a message, why doesn't it turn off the paths for these roads?
-		std::string roadsMsg = "Ped roads have been switched off for: "
-		+ std::string("leftBottomX: ") + std::to_string(leftBottom->x) 
-		+ " leftBottomY: " + std::to_string(leftBottom->y)
-		+ " leftBottomZ: " + std::to_string(leftBottom->z)
-		
-		+ " rightTopX: " + std::to_string(rightTop->x)
-		+ " rightTopY: " + std::to_string(rightTop->y)
-		+ " rightTopZ: " + std::to_string(rightTop->z);
+		if(infZ > supZ)
+		{
+			infZ = supZ;
+			supZ = infZ;
+		}
+	
+		// if (!leftBottom || !rightTop)
+    	// {
+    	// 	return luaL_error(L, "world.switch_ped_roads_off expects two vector tables with x, y and z values." );
+		// }
 
-		// LogFunctions::LogInfo(Defines::luaLogFile, roadsMsg);
-		//
+		ThePaths.SwitchPedRoadsOffInArea(infX, supX, infY, supY, infZ, supZ, true);
 
 		return 0;
 	}
@@ -502,6 +510,8 @@ extern "C"
 	/**
 	 * Switch the ped roads on in the area.
 	 * 
+	 * TODO Fix this, match up with turn off function later.
+	 * 
 	 * Usage:
 	 * world.switch_ped_roads_on({x = 25, y = 25, z = 25}, {x = 50, y = 50, z = 50})
 	 */
@@ -510,6 +520,8 @@ extern "C"
 		auto leftBottom = luaGetCVector(L, 1);
 		auto rightTop = luaGetCVector(L, 2);
 	
+		// Replicated setup from COMMAND_SWITCH_ROADS_OFF.
+
 		if (!leftBottom || !rightTop)
     	{
     		return luaL_error(L, "world.switch_ped_roads_on expects two vector tables with x, y and z values." );
@@ -759,8 +771,6 @@ extern "C"
 	/**
 	 * Fade the camera in or out with a specific time.
 	 * 
-	 * TODO Fix this to work, I don't think it will work without a wait timer.
-	 * 
 	 * Usage:
 	 * world.fade_camera(2.0, camera_enums.eFadeDirection.FADE_OUT)
 	 * 
@@ -790,6 +800,19 @@ extern "C"
 		return 0;
 	}
 
+	/**
+	 * Check if the camera is fading in or out.
+	 * 
+	 * Usage:
+	 * world.get_fading_status()
+	 */
+	int getFadingStatusLua(lua_State *L)
+	{
+		bool isCameraFading = TheCamera.GetFading();
+		pushBooleanToLua(L, isCameraFading);
+		return 1;
+	}
+
 	/**
 	 * Setup a target marker for loading the json save file in lua.
 	 * 
@@ -898,13 +921,36 @@ extern "C"
 		int i;
 
 		const char *gxtLabel = lua_tostring(L, 1);
+		auto dayOrNight = luaL_checkinteger(L, 2);
+
+		auto density = luaL_checkinteger(L, 3);
+		auto cuban = luaL_checkinteger(L, 4);
+		
+		// 
+		auto haitian = luaL_checkinteger(L, 5);
+		auto street = luaL_checkinteger(L, 6);
+		//
+
+		bool diaz = luaL_checkinteger(L, 7);
+
+		auto security = luaL_checkinteger(L, 8);
+		auto biker = luaL_checkinteger(L, 9);
+		auto player = luaL_checkinteger(L, 10);
+		auto golfer = luaL_checkinteger(L, 11);
+		auto gang9 = luaL_checkinteger(L, 12);
+		auto cop = luaL_checkinteger(L, 13);
+		//
 
 
 		// Set the gang densities.
 		// TODO Implement this part.
-		// for (i = 0; i < NUM_GANGS; i++)
-			// break;
+		for (i = 0; i < NUM_GANGS; i++)
+			gangDensities[i] = lua_tonumber(L, i + 2);
 			// gangDensities[i] = ScriptParams[i + 2];
+			// break;
+
+		// for (i = 0; i < )
+			
 
 		int16 zone = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
 
@@ -918,10 +964,16 @@ extern "C"
 		}
 
 		// TODO Figure this out.
-		// while (zone >= 0) {
-		// 	CTheZones::SetZoneCarInfo(zone, ScriptParams[0], ScriptParams[1], ScriptParams[11], gangDensities);
-		// 	zone = CTheZones::FindNextZoneByLabelAndReturnIndex(label, ZONE_INFO);
-		// }
+		// TODO Test this, i doubt it'll work or do anything.
+		// I may make another function that replicates the ped one, it just takes all the parameters normally.
+		// Instead of trying to do this array shit..
+		while (zone >= 0) {
+			// CTheZones::SetZoneCarInfo(zone, dayOrNight, ScriptParams[1], ScriptParams[11], gangDensities);
+			// CTheZones::SetZoneCarInfo(zone, dayOrNight, density, ScriptParams[11], gangDensities);
+			// for (int i = 0; i < )
+			CTheZones::SetZoneCarInfo(zone, dayOrNight, density, cop, gangDensities);
+			zone = CTheZones::FindNextZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
+		}
 
 		return 0;
 	}
@@ -946,13 +998,53 @@ extern "C"
 
 		const char *gxtLabel = lua_tostring(L, 1);
 		auto dayOrNight = luaL_checkinteger(L, 2);
+		auto normal = luaL_checkinteger(L, 3);
+		auto poor = luaL_checkinteger(L, 4);
+		
+		// 
+		auto rich = luaL_checkinteger(L, 5);
+		auto exex = luaL_checkinteger(L, 6);
+		//
+
+		bool worker = luaL_checkinteger(L, 7);
+
+		auto big = luaL_checkinteger(L, 8);
+		auto taxi = luaL_checkinteger(L, 9);
+		auto moped = luaL_checkinteger(L, 10);
+		auto motorbike = luaL_checkinteger(L, 11);
+		auto leisureBoat = luaL_checkinteger(L, 12);
+		auto workerBoat = luaL_checkinteger(L, 13);
+		//
+
+		// I found this, it might help out with setting this up.
+		// https://github.com/neomutt/lua-samples/blob/main/callback.c
+
+		// Get the number of arguments
+		int arguments = lua_gettop(L);
+
+		// I doubt these below will work.. But maybe.
+		// Start this at three? Not sure..
+		// Subtract two from the arguments, should get the amount of vehicle densities.
+		// for (int i = 3; i < arguments - 2; i++)
+		// {
+		// 	carDensities[i] = lua_tonumber(L, i);
+		// }
+
+		// for (int i = arguments - 2; i < arguments; i ++)
+		// {
+		// 	boatDensities[i] = lua_tonumber(L, i);
+		// }
+
+
 
 		// TODO Figure this part out.
-		// for (i = 0; i < CCarCtrl::NUM_CAR_CLASSES; i++)
+		for (i = 0; i < CCarCtrl::NUM_CAR_CLASSES; i++)
 			// carDensities[i] = ScriptParams[i + 1];
+			carDensities[i] = lua_tonumber(L, i + 1);
 
-		// for (i = 0; i < CCarCtrl::NUM_BOAT_CLASSES; i++)
+		for (i = 0; i < CCarCtrl::NUM_BOAT_CLASSES; i++)
 			// boatDensities[i] = ScriptParams[i + 1 + CCarCtrl::NUM_CAR_CLASSES];
+			boatDensities[i] = lua_tonumber(L, i + 1 + CCarCtrl::NUM_CAR_CLASSES);
 		//
 
 		int16 zone = CTheZones::FindZoneByLabelAndReturnIndex(const_cast<char*>(gxtLabel), ZONE_INFO);
@@ -1117,6 +1209,43 @@ extern "C"
 		return 0;
 	}
 
+	/**
+	 * Draw a sphere in an area
+	 * 
+	 * 
+	 * Usage:
+	 * world.draw_sphere(id, {x = 25, y = 25, z = 25}, 5.0)
+	 * 
+	 */
+	int drawSphereLua(lua_State *L)
+	{
+		auto sphereId = luaL_checknumber(L, 1);
+		auto spherePos = luaGetCVector(L, 2);
+		auto sphereRadius = luaL_checknumber(L, 3);
+
+		// It works to make a random one.
+		C3dMarkers::PlaceMarkerSet(sphereId, MARKERTYPE_CYLINDER, *spherePos, sphereRadius,
+			SPHERE_MARKER_R, SPHERE_MARKER_G, SPHERE_MARKER_B, SPHERE_MARKER_A,
+			SPHERE_MARKER_PULSE_PERIOD, SPHERE_MARKER_PULSE_FRACTION, 0);
+
+		return 0;
+	}
+
+	/**
+	 * Toggle the trash and stuff on the ground
+	 * 
+	 * Usage:
+	 * world.switch_rubbish(true)
+	 * 
+	 */
+	int switchRubbishLua(lua_State *L)
+	{
+		bool state = static_cast<float>(lua_toboolean(L, 2));
+		CRubbish::SetVisibility(state);
+
+		return 0;
+	}
+
 	/**
 	 * Get a random cop in the area.
 	 * TODO Implement this.
diff --git a/src/extras/lua/lua_world.h b/src/extras/lua/lua_world.h
index 0b664b76..39ac5597 100644
--- a/src/extras/lua/lua_world.h
+++ b/src/extras/lua/lua_world.h
@@ -50,6 +50,7 @@ extern "C" {
 
 	// Fading
 	extern int fadeCameraLua(lua_State *L);
+	extern int getFadingStatusLua(lua_State *L);
 
 	// Target marker
 	extern int setupTargetMarkerLua(lua_State *L);
@@ -64,6 +65,9 @@ extern "C" {
 	extern int setZoneCivilianCarInfoLua(lua_State *L);
 	extern int setZoneGroupLua(lua_State *L);
 
+	extern int drawSphereLua(lua_State *L);
+	extern int switchRubbishLua(lua_State *L);
+
 
 #ifdef __cplusplus
 }
diff --git a/src/weapons/Weapon.cpp b/src/weapons/Weapon.cpp
index 97142027..910c0489 100644
--- a/src/weapons/Weapon.cpp
+++ b/src/weapons/Weapon.cpp
@@ -474,11 +474,20 @@ CWeapon::FireFromCar(CVehicle *shooter, bool left, bool right)
 	{
 		DMAudio.PlayOneShot(shooter->m_audioEntityId, SOUND_WEAPON_SHOT_FIRED, 0.0f);
 
-		if ( m_nAmmoInClip > 0 )
+#ifdef DISABLE_GAME_SCRIPTS
+			if (m_nAmmoInClip > 0 && !gbInfiniteAmmoCheat)
+#else
+			if  (m_nAmmoInClip > 0 )
+#endif // DISABLE_GAME_SCRIPTS - Infinite ammo cheat.
 			m_nAmmoInClip--;
 
-		if ( m_nAmmoTotal < 25000 && m_nAmmoTotal > 0 && (!shooter || shooter->GetStatus() != STATUS_PLAYER || CStats::GetPercentageProgress() < 100.f))
-			m_nAmmoTotal--;
+			if ( m_nAmmoTotal < 25000 && m_nAmmoTotal > 0 && (!shooter || shooter->GetStatus() != STATUS_PLAYER || CStats::GetPercentageProgress() < 100.f
+#ifdef DISABLE_GAME_SCRIPTS
+				&& !gbInfiniteAmmoCheat
+#endif // DISABLE_GAME_SCRIPTS - Infinite ammo cheat.
+				))
+				m_nAmmoTotal--;
+			//---------------------
 
 		m_eWeaponState = WEAPONSTATE_FIRING;
 
-- 
2.43.0

